background.js

```diff
@@ -10,21 +21,152 @@ chrome.runtime.onMessage.addListener(async (message) => {
+      'enable_auto_archive',
+      'match_urls',
+      'exclude_urls',
+    ]);
+
+    console.debug(`[Auto-Archive Debug] Settings: enable_auto_archive=${enable_auto_archive}, match_urls="${match_urls}", exclude_urls="${exclude_urls}"`);
+
+    if (!enable_auto_archive || !match_urls || match_urls.trim() === '') {
+      console.debug('[Auto-Archive Debug] Auto-archiving disabled or match pattern empty');
+      return false;
+    }
+
+    const matchPattern = new RegExp(match_urls);
+    const matches = matchPattern.test(url);
+    console.debug(`[Auto-Archive Debug] URL match test: ${matches} (pattern: ${matchPattern})`);
+
+    if (!matches) {
+      return false;
+    }
+
+    if (exclude_urls && exclude_urls.trim() !== '') {
+      try {
+        const excludePattern = new RegExp(exclude_urls);
+        const excluded = excludePattern.test(url);
+        console.debug(`[Auto-Archive Debug] URL exclude test: ${excluded} (pattern: ${excludePattern})`);
+
+        if (excluded) {
+          return false;
+        }
+      } catch (error) {
+        console.error('Invalid exclude pattern:', error);
+      }
+    }
+
+    console.debug(`[Auto-Archive Debug] URL ${url} should be archived: TRUE`);
+    return true;
+  } catch (error) {
+    console.error('Error checking auto-archive patterns:', error);
+    return false;
+  }
+}
+
+// Global reference to the listener so we can remove it properly
+let tabUpdateListener = null;
+
+async function setupAutoArchiving() {
+  console.debug('[Auto-Archive Debug] Setting up auto-archiving...');
+  const { enable_auto_archive } = await chrome.storage.local.get(['enable_auto_archive']);
+  console.debug(`[Auto-Archive Debug] enable_auto_archive setting: ${enable_auto_archive}`);
+
+  if (tabUpdateListener) {
+    try {
+      console.debug('[Auto-Archive Debug] Removing existing tab update listener');
+      chrome.tabs.onUpdated.removeListener(tabUpdateListener);
+      tabUpdateListener = null;
+    } catch (error) {
+      console.error('Error removing listener:', error);
+    }
+  }
+
+  if (enable_auto_archive) {
+    console.debug('[Auto-Archive Debug] Auto-archive is enabled, checking permissions');
+    const hasPermission = await chrome.permissions.contains({ permissions: ['tabs'] });
+    console.debug(`[Auto-Archive Debug] Has tabs permission: ${hasPermission}`);
+
+    if (hasPermission) {
+      tabUpdateListener = async (tabId, changeInfo, tab) => {
+        console.debug(`[Auto-Archive Debug] Tab updated - tabId: ${tabId}, status: ${changeInfo.status}, url: ${tab?.url}`);
+
+        // Only process when the page has completed loading
+        if (changeInfo.status === 'complete' && tab.url) {
+          console.debug(`[Auto-Archive Debug] Tab load complete, checking if URL should be archived: ${tab.url}`);
+
+          const shouldArchive = await shouldAutoArchive(tab.url);
+          console.debug(`[Auto-Archive Debug] shouldAutoArchive result: ${shouldArchive}`);
+
```

Comment:
needs to check that the URL is not already archived (locally) otherwise there will be tons of duplicates